bookkeeper-server/src/test/java/org/apache/bookkeeper/proto/ReadEntryProcessorTest.java

```diff
@@ -0,0 +1,136 @@
+        when(bookie.fenceLedger(anyLong(), any())).thenReturn(fenceResult);
+
+        ChannelPromise promise = new DefaultChannelPromise(channel);
+        AtomicReference<Object> writtenObject = new AtomicReference<>();
+        CountDownLatch latch = new CountDownLatch(1);
+        doAnswer(invocationOnMock -> {
+            writtenObject.set(invocationOnMock.getArgument(0));
+            latch.countDown();
+            return promise;
+        }).when(channel).writeAndFlush(any(Response.class), any());
+
+        ExecutorService service = Executors.newCachedThreadPool();
+        long ledgerId = System.currentTimeMillis();
+        ReadRequest request = new ReadRequest(BookieProtocol.CURRENT_PROTOCOL_VERSION, ledgerId,
+                1, BookieProtocol.FLAG_DO_FENCING, new byte[]{});
+        ReadEntryProcessor processor = ReadEntryProcessor.create(request, channel, requestProcessor, service);
+        processor.run();
+
+        fenceResult.set(true);
+        latch.await();
+        verify(channel, times(1)).writeAndFlush(any(Response.class), any());
+
+        assertTrue(writtenObject.get() instanceof Response);
+        Response response = (Response) writtenObject.get();
+        assertEquals(1, response.getEntryId());
+        assertEquals(ledgerId, response.getLedgerId());
+        assertEquals(READENTRY, response.getOpCode());
+        assertEquals(BookieProtocol.EOK, response.getErrorCode());
+    }
+
+    @Test
+    public void testFenceRequestWithoutThreadPool() throws Exception {
+        SettableFuture<Boolean> fenceResult = SettableFuture.create();
+        when(bookie.fenceLedger(anyLong(), any())).thenReturn(fenceResult);
+
+        ChannelPromise promise = new DefaultChannelPromise(channel);
+        AtomicReference<Object> writtenObject = new AtomicReference<>();
+        CountDownLatch latch = new CountDownLatch(1);
+        doAnswer(invocationOnMock -> {
+            writtenObject.set(invocationOnMock.getArgument(0));
+            latch.countDown();
+            return promise;
+        }).when(channel).writeAndFlush(any(Response.class), any());
+
+        long ledgerId = System.currentTimeMillis();
+        ReadRequest request = new ReadRequest(BookieProtocol.CURRENT_PROTOCOL_VERSION, ledgerId,
+                1, BookieProtocol.FLAG_DO_FENCING, new byte[]{});
+        ReadEntryProcessor processor = ReadEntryProcessor.create(request, channel, requestProcessor, null);
+        fenceResult.set(true);
+        processor.run();
+
+        latch.await();
+        verify(channel, times(1)).writeAndFlush(any(Response.class), any());
+
+        assertTrue(writtenObject.get() instanceof Response);
+        Response response = (Response) writtenObject.get();
+        assertEquals(1, response.getEntryId());
+        assertEquals(ledgerId, response.getLedgerId());
+        assertEquals(READENTRY, response.getOpCode());
+        assertEquals(BookieProtocol.EOK, response.getErrorCode());
```

Comment:
Very good.
I am really sorry I missed one comment.
Please a test in which the response contains BookieProtocol.EIO error code.

Reply:
Please review again.